crates/store/src/state.rs

```diff
@@ -236,17 +241,10 @@ impl State {
         // in-memory write lock. This requires the DB update to run concurrently, so a new task is
         // spawned.
         let db = self.db.clone();
-        let handle = tokio::spawn(async move {
-            db.apply_block(
-                allow_acquire,
-                acquire_done,
-                block.header(),
-                notes,
-                block.created_nullifiers().to_vec(),
-                block.updated_accounts().to_vec(),
-            )
-            .await
-        });
+        let handle =
+            tokio::spawn(
+                async move { db.apply_block(allow_acquire, acquire_done, block, notes).await },
+            );
```

Comment:
I moved `blockstore` into `State`. Regarding simplification, if you mean removing `tokio::spawn` and communication using channels here, I encourage that. I also proposed to remove this construction long time ago. :) I don't know what benefits we get from this now, since once transaction commit fails, we still put in-memory storage into inconsistent state with the database.

Reply:
> I don't know what benefits we get from this now, since once transaction commit fails, we still put in-memory storage into inconsistent state with the database.

I don't think that's the case: if transaction commit fails, the in-memory data structures are not updated - i.e., the `mem::replace` statements a few lines down are not executed (or at least, that's the intent).

In general, this mechanism is used to support the following process:

1. We start the database transaction and update the database. While this is going on, reads against the database are allowed as they read the previous state. So, the store can still serve all requests.
2. Once the data has been written to the database, we lock the in-memory structures for update. At this point, the store blocks on all requests.
3. We update in-memory structures. While this is going on, the store cannot serve any requests - so, it is critical to make this section as fast as possible.
4. We commit the database transaction and release the lock on in-memory structures.

Because database transaction and in-memory structures live in different places, we need this channel - but I am wondering if there is some way we can refactor the structure to make the code simpler.

Reply:
Thank you for explanation, now it's much clearer for me. According to the explanation, current implementation still has bugs (3 actually):
1. When we lock in-memory data, we don't wait until commit is succeed and start to update in-memory. In a case of failed commit, in-memory data will be updated.
2. Once we finish update in-memory data, we unlock it instantly without waiting for finishing of database update.
3. Error result of `db.apply_block` in `tokio::spawn` is swallowed.

Reply:
> 1. When we lock in-memory data, we don't wait until commit is succeed and start to update in-memory. Even in a case of failed commit, in-memory data will be updated. By the way, we have a comment about this below.

I'm not sure that's correct. Currently, we make copies of the data and start updating them (this is very inefficient, and we'll need to improve it in the future, but that's how it works now). Then, once the transaction is committed, we do `mem::replace` to swap out the updated copies.

> 2. Once we finish update in-memory data, we unlock it instantly without waiting for finishing of database update.

I think this is similar to the point above: in-memory updates don't actually happen until `mem::replace` calls, and these happen after the database transaction is committed.

> 3. Error result of `db.apply_block` in `tokio::spawn` only logged, but not re-raised or handled.
> 
> The simplest solution would be just moving all these `mem::replace`s into `Ok(Ok(()))` branch of result handling below (line 283).

This part may be correct - but it is difficult for me to tell how error propagation works here. So, probably another good reason to try to refactor the code (and document it more) to make it super clear what's happening.

Reply:
@bobbinth, it actually works differently. 

Please look at these lines: https://github.com/0xPolygonMiden/miden-node/blob/816d668f03f3652183a1f16a65c1d6c59faf2cf9/crates/store/src/db/mod.rs#L306-L309

Here our spawned task sends notification that it's a time for acquiring the lock and then waits for the response. Response is sent by the caller task just after locking (the caller awaits for the notification first, then acquires the lock and sends another notification, after that it immediately starts to update in-memory data - without waiting for commit): https://github.com/0xPolygonMiden/miden-node/blob/816d668f03f3652183a1f16a65c1d6c59faf2cf9/crates/store/src/state.rs#L253-L269

Reply:
Hmmm - I'll need to look into this more. Let's create an issue (and link this discussion there). We can address it as a part of the next milestone.